src/smt.rs

```diff
@@ -954,97 +965,116 @@ impl SparseMerkleTree {
-                        .map_err(|err| SMTError::Unknown(err.to_string()))?;
-                    subtree
-                },
-                NodeKind::Empty => self.get_subtree(db, &info.current_node.hash)?,
-                NodeKind::Leaf => {
-                    SubTree::from_data(vec![0], vec![info.current_node], self.hasher)?
-                },
-                _ => {
-                    return Err(SMTError::Unknown(String::from("invalid node type")));
-                },
-            };
-            if info.key_bins.len() != 1 || info.value_bins.len() != 1 {
-                return Err(SMTError::Unknown(String::from("invalid key/value length")));
-            }
-            let new_subtree = self.update_subtree(
-                db,
-                info.key_bins[0].clone(),
-                info.value_bins[0].clone(),
-                &btm_subtree,
-                info.height + info.h as u32,
-            )?;
-            if new_subtree.nodes.len() == 1 {
-                return Ok((vec![new_subtree.nodes[0].clone()], vec![info.h]));
-            }
-            let new_branch = Node::new_stub(new_subtree.root.as_slice());
+        Ok((vec![], vec![]))
+    }
 
-            return Ok((vec![new_branch], vec![info.h]));
+    fn update_same_height(
+        &mut self,
+        db: &mut impl DB,
+        info: &UpdateNodeInfo,
+    ) -> Result<(Vec<Node>, Vec<u8>), SMTError> {
+        let btm_subtree = match info.current_node.kind {
+            NodeKind::Stub => {
+                let subtree = self.get_subtree(db, &info.current_node.hash)?;
+                db.del(info.current_node.hash.clone())
+                    .map_err(|err| SMTError::Unknown(err.to_string()))?;
+                subtree
+            },
+            NodeKind::Empty => self.get_subtree(db, &info.current_node.hash)?,
+            NodeKind::Leaf => {
+                SubTree::from_data(vec![0], vec![info.current_node.clone()], self.hasher)?
+            },
+            _ => {
+                return Err(SMTError::Unknown(String::from("invalid node type")));
+            },
+        };
+        if info.key_bins.len() != 1 || info.value_bins.len() != 1 {
+            return Err(SMTError::Unknown(String::from("invalid key/value length")));
+        }
+        let new_subtree = self.update_subtree(
+            db,
+            info.key_bins[0].clone(),
+            info.value_bins[0].clone(),
+            &btm_subtree,
+            info.height + info.h as u32,
+        )?;
+        if new_subtree.nodes.len() == 1 {
+            return Ok((vec![new_subtree.nodes[0].clone()], vec![info.h]));
         }
+        let new_branch = Node::new_stub(new_subtree.root.as_slice());
 
-        let (left_node, right_node) = match info.current_node.kind {
-            NodeKind::Empty => (Node::new_empty(), Node::new_empty()),
+        Ok((vec![new_branch], vec![info.h]))
+    }
+
+    fn left_right_nodes(&self, info: &UpdateNodeInfo) -> Result<(Node, Node), SMTError> {
```

Comment:
this function is returning next branch to compute based on the current node.
for example, if current node is empty, both left/right will be updated as empty, but if it's leaf, one of them is leaf and another one will be updated as empty

Reply:
What about `next_branch_action` ?

Reply:
I think `left_right_nodes` would be better, so let's keep it as we discussed